### PR TITLE
Code is now compatible with ME v5.0.0.2122

### PR DESCRIPTION
**Pull request by nseam**
Fixes: #29, #57, #58 and #59

### CODE REVIEW

File: Chart.mqh [hunk 1]
```diff
@@ -824,28 +824,28 @@ class Chart : public Market {
...
-      return fabs(GetClose(_bar) - GetOpen(_bar)) / GetPointsPerPip();
+      return fabs(Chart::GetClose((int)_bar) - Chart::GetOpen((uint)_bar)) / GetPointsPerPip();
     }
 
     /**
      * Returns bar's head size in pips.
      */
     double GetBarHeadSize(uint _bar) {
-      return (GetHigh(_bar) - fmax(GetClose(_bar), GetOpen(_bar))) / GetPointsPerPip();
+      return (GetHigh(_bar) - fmax(Chart::GetClose((int)_bar), Chart::GetOpen((uint)_bar))) / GetPointsPerPip();
```
Comment (kenorb) [inline Chart.mqh]: Can we use consistent casting, either `uint` or `int`? Same below.

File: Chart.mqh [hunk 2]
```diff
@@ -824,28 +824,28 @@ class Chart : public Market {
...
     }
 
     /**
      * Returns bar's tail size in pips.
      */
     double GetBarTailSize(uint _bar) {
-      return (fmin(GetClose(_bar), GetOpen(_bar)) - GetLow(_bar)) / GetPointsPerPip();
+      return (fmin(Chart::GetClose((int)_bar), Chart::GetOpen(_bar)) - GetLow(_bar)) / GetPointsPerPip();
```
Comment (kenorb) [inline Chart.mqh]: For consistency, we should use `Market::GetPointsPerPip` instead?

File: Chart.mqh [hunk 3]
```diff
@@ -693,10 +693,10 @@ class Chart : public Market {
      * Calculates pivot points in different systems.
      */
     static void CalcPivotPoints(string _symbol, ENUM_TIMEFRAMES _tf, ENUM_PP_METHOD _method, double &PP, double &S1, double &S2, double &S3, double &S4, double &R1, double &R2, double &R3, double &R4) {
-      double _open   = iOpen(_symbol, _tf, 1);
-      double _high   = iHigh(_symbol, _tf, 1);
-      double _low    = iLow(_symbol, _tf, 1);
-      double _close  = iClose(_symbol, _tf, 1);
+      double _open   = ::iOpen(_symbol, _tf, 1);
```
Comment (kenorb) [inline Chart.mqh]: You should use `Chart::iOpen` instead, as `::iOpen` won't exist in MQL5?

File: Chart.mqh [hunk 4]
```diff
@@ -654,16 +654,16 @@ class Chart : public Market {
       // Nearest time.
       nBars = iBars(NULL, (ENUM_TIMEFRAMES) TimePr);
       for (i = 0; i < nBars;i++) {
-        if (iOpen(NULL,TimePr, i) >= 0.000001) {
-          if (iTime(NULL, TimePr, i) >= modeling_start_time)
+        if (::iOpen(NULL,TimePr, i) >= 0.000001) {
```
Comment (kenorb) [inline Chart.mqh]: Use `Chart::iOpen` to make it work for MT5.

File: Chart.mqh [hunk 5]
```diff
@@ -654,16 +654,16 @@ class Chart : public Market {
...
+          if (::iTime(NULL, TimePr, i) >= modeling_start_time)
```
Comment (kenorb) [inline Chart.mqh]: Use `Chart:: iTime` to make it work for MT5.

File: Chart.mqh (hunk 1 above)
Comment (nseam) [reply Chart.mqh]: Fixed

File: Chart.mqh (hunk 2 above)
Comment (nseam) [reply Chart.mqh]: Fixed

File: Chart.mqh (hunk 4 above)
Comment (nseam) [reply Chart.mqh]: Fixed

File: Chart.mqh (hunk 3 above)
Comment (nseam) [reply Chart.mqh]: Fixed

File: Chart.mqh (hunk 5 above)
Comment (nseam) [reply Chart.mqh]: Fixed

File: Condition.mqh [hunk 6]
```diff
@@ -35,6 +35,79 @@
...
+       C_ATR_BUY_SELL        =  5, // ATR on buy/sell
+    C_AWESOME_BUY_SELL    =  6, // Awesome on buy/sell
+    C_BANDS_BUY_SELL      =  7, // Bands on buy/sell
+    C_BEARSPOWER_BUY_SELL =  8, // BearsPower on buy/sell
+    C_BULLSPOWER_BUY_SELL = 40, // BullsPower on buy/sell
+    C_BWMFI_BUY_SELL      = 10, // BWMFI on buy/sell
+    C_CCI_BUY_SELL        = 11, // CCI on buy/sell
+    C_DEMARKER_BUY_SELL   = 12, // DeMarker on buy/sell
+    C_ENVELOPES_BUY_SELL  = 13, // Envelopes on buy/sell
+       C_FORCE_BUY_SELL      = 14, // Force on buy/sell
```
Comment (kenorb) [inline Condition.mqh]: Wrong indentation.

File: Condition.mqh [hunk 7]
```diff
@@ -35,6 +35,79 @@
...
+// Define market event conditions.
+#ifndef MARKET_EVENT_ENUM
+  #define MARKET_EVENT_ENUM
+  enum ENUM_MARKET_EVENT {
+    C_EVENT_NONE          =  0, // None
+    C_AC_BUY_SELL         =  1, // AC on buy/sell
+    C_AD_BUY_SELL         =  2, // AD on buy/sell
+    C_ADX_BUY_SELL        =  3, // ADX on buy/sell
+    C_ALLIGATOR_BUY_SELL  =  4, // Alligator on buy/sell
+       C_ATR_BUY_SELL        =  5, // ATR on buy/sell
```
Comment (kenorb) [inline Condition.mqh]: Wrong indentation.

File: Condition.mqh [hunk 8]
```diff
@@ -35,6 +35,79 @@
...
+    C_CUSTOM4_MARKET_COND = 37, // Custom 4 market condition
+    C_CUSTOM5_MARKET_COND = 38, // Custom 5 market condition
+    C_CUSTOM6_MARKET_COND = 39, // Custom 6 market condition
+  #endif
+  };
+#endif
+
+// Define market conditions.
+enum ENUM_MARKET_CONDITION {
+ C_MARKET_NONE        = 00, // None (false).
```
Comment (kenorb) [inline Condition.mqh]: Use 2 spaces for indentation.

File: Condition.mqh [hunk 9]
```diff
@@ -35,6 +35,79 @@
 // Define an assert macros.
 #define METHOD(method, no) ((method & (1<<no)) == 1<<no)
 
+// Define market event conditions.
+#ifndef MARKET_EVENT_ENUM
+  #define MARKET_EVENT_ENUM
+  enum ENUM_MARKET_EVENT {
+    C_EVENT_NONE          =  0, // None
```
Comment (kenorb) [inline Condition.mqh]: Use 2 spaces for indentation.

File: Chart.mqh [hunk 10]
```diff
@@ -95,26 +95,34 @@ const ENUM_TIMEFRAMES arr_tf[TFS] = {
...
-  } cparams;
-  // Struct for storing OHLC values.
-  struct OHLC {
-    datetime time;
-    double open, high, low, close;
-  } ohlc_saves[];
+  
+  public:
+  
+   ChartParams cparams;
```
Comment (kenorb) [inline Chart.mqh]: Use 2 spaces for indentation.

File: Chart.mqh [hunk 11]
```diff
@@ -95,26 +95,34 @@ const ENUM_TIMEFRAMES arr_tf[TFS] = {
...
 };
 
+// Struct for storing OHLC values.
+struct OHLC {
+ datetime time;
+ double open, high, low, close;
+};
+
+struct ChartParams {
+ ENUM_TIMEFRAMES tf;
```
Comment (kenorb) [inline Chart.mqh]: Use 2 spaces for indentation.

File: Chart.mqh [hunk 12]
```diff
@@ -95,26 +95,34 @@ const ENUM_TIMEFRAMES arr_tf[TFS] = {
   PERIOD_D1, PERIOD_W1, PERIOD_MN1
 };
 
+// Struct for storing OHLC values.
+struct OHLC {
+ datetime time;
```
Comment (kenorb) [inline Chart.mqh]: Use 2 spaces for indentation.

**Pull request merged**